src/tokens/Source.js

```diff
@@ -76,34 +107,45 @@ export class Source {
             }
         }
 
-        let c = this.content[i]
+        let c = this.getChar(i)
 
         while (isWordChar(c)) {
             chars.push(c)
             i += 1
-            c = this.content[i]
+            c = this.getChar(i)
         }
 
         return chars.join("")
     }
 
+    /**
+     * @type{[number, number][] | undefined}
+     */
+    lineLengths
```

Comment:
I would change the name of this property to `_lineLengths` and make it `@private`, then change its type to `Option<[number, number][]>` and initialize it explicitly in the constructor using `None` imported from @helios-lang/type-utils.

The `calcLineLengths()` function I would change to `get lineLengths()`, which returns either the cached result from the private property, or does the calculation (assigning to the private property in the process).

That way any function that calls `this.lineLengths` doesn't have to worry about the caching.

Reply:
They're actually line-ending locations; renamed.

The first number in the tuple wasn't actually used/needed, so I changed the type to `number[]` too.